src/Tools/dotnet-monitor/CollectionRules/CollectionRulePipeline.cs

```diff
@@ -65,14 +75,16 @@ protected override async Task OnRun(CancellationToken token)
 
             CancellationToken linkedToken = linkedCancellationSource.Token;
 
-            TimeSpan? actionCountWindowDuration = _context.Options.Limits?.ActionCountSlidingWindowDuration;
-            int actionCountLimit = (_context.Options.Limits?.ActionCount).GetValueOrDefault(CollectionRuleLimitsOptionsDefaults.ActionCount);
-            Queue<DateTime> executionTimestamps = new(actionCountLimit);
+            _actionCountWindowDuration = Context.Options.Limits?.ActionCountSlidingWindowDuration;
+            _actionCountLimit = (Context.Options.Limits?.ActionCount).GetValueOrDefault(CollectionRuleLimitsOptionsDefaults.ActionCount);
+            _executionTimestamps = new Queue<DateTime>(_actionCountLimit);
```

Comment:
I would like to see at least `_extensionTimestamps` and `AllExecutionTimestamps` moved to `CollectionRuleStateHolder` as well as move the `DequeOldTimestamp` and `CheckForThottling` logic moved there as well. Then expose something like `CanExecuteActions(ISystemClock)` and `ActionExecutionComplete(bool success)` methods. These would be called by where you currently call `CheckForThrottling`. The `BeginThrottled`, `EndThrottled`, `ActionExecutionFailed`, `ActionExecutionSucceeded` can be removed and become implementation details of two new methods.

You might have to initialize the state holder with the collection rule limits.

The benefit of this is that all of the execution state is now in the state holder and doesn't require calling static methods on `CollectionRulePipeline`. The webapi can effectively get a copy of the state holder and not have to ever touch the pipeline again.